src/cpp/core/include/core/FileLock.hpp

```diff
@@ -17,30 +17,90 @@
 #define CORE_FILE_LOCK_HPP
 
 #include <boost/utility.hpp>
+#include <boost/shared_ptr.hpp>
 #include <boost/scoped_ptr.hpp>
+#include <boost/asio.hpp>
+#include <boost/function.hpp>
+#include <boost/bind.hpp>
 
 namespace rstudio {
 namespace core {
 
+namespace locks {
+void initialize();
+} // namespace locks
+
 class Error;
 class FilePath;
 
 class FileLock : boost::noncopyable
 {
 public:
-   static bool isLocked(const FilePath& lockFilePath);
+   
+   enum LockType { LOCKTYPE_ADVISORY, LOCKTYPE_LINKBASED };
+   
+   // control what kind of locks are generated by 'create()' method
+   static void setLockType(LockType type) { type_ = type; }
+   static LockType getLockType() { return type_; }
+   static boost::shared_ptr<FileLock> create();
+   
+   // control lock timeout
+   static void setLockTimeoutInterval(boost::posix_time::seconds timeoutInterval) { timeoutInterval_ = timeoutInterval; }
```

Comment:
Maybe a single `initialize` method where we pass in all of the options? I'm thinking that this makes it much more clear that the `set` methods should only be called at startup before multiple threads are in play (since they don't have synchronization protecting the get/set). BTW I think it's perfectly fine and even preferable not to synchronize values that are only set at startup it's just nice to make those semantics clear in the interface.